### DISCUSSION TITLE
Filter based on deep (nested level) item's Many To Any Field

### DISCUSSION THREAD

**Discussion by Disura-Randunu** [directus/directus#8061] [Archive]
When we filter, is it possible to compare a top level field value to a deep or nested (Can be also multi level) field's value in the SDK?

**Comment by rijkvanzanten** [comment]
What do you mean by "compare" the value?

**Reply by Disura-Randunu** [reply]
Let's say we have a field in top-level called 'id'.  How can we filter out the item which has the same value of 'id' in a nested level field?

**Reply by Disura-Randunu** [reply]
I think my explanation is not clear. Sorry about that. Anyway, this is the error I get when I try to compare the many to any field value in query:

> You have to provide a collection scope when filtering on a many-to-any item

How can we solve this error?

Query I tried:
```
second_interactor: {
 interactors_id: {
   interactor: {
      item: {
         project_id: {
            _eq: '29540798'
         }
      }
    }
  }
}                                                                
```

Sample output item without any filter:
```
{
            "id": 1,
            "valid_from": null,
            "valid_to": null,
            "second_interactor": [
                {
                    "id": 1,
                    "interactors_id": {
                        "id": 2,
                        "name": "Metric Dashboard Gitlab",
                        "interactor_type": {
                            "id": 3,
                            "type": "gitlab"
                        },
                        "interactor": [
                            {
                                "id": 2,
                                "collection": "gitlab",
                                "interactors_id": {
                                    "id": 2,
                                    "interactor_type": 3,
                                    "name": "Metric Dashboard Gitlab",
                                    "interactor": [
                                        2
                                    ]
                                },
                                "item": {
                                    "id": 1,
                                    "project_id": "29540798"
                                }
                            }
                        ]
                    }
                }
            ]
        },
```
Item is not a collection but a field that gets created in the intermediate table. So how can we refer to the actual item which can be from any table since it's a Many to Any relation?

**Reply by Disura-Randunu** [reply]
I found a discord chat similar to this issue, but it uses the API. Something like this: "filter": { "intermediateTable:ActualTable.id": 3}. I tried this in the SDK for my problem like below:
```
filter:{
 second_interactor: {
    interactors_id: {
       interactor: {
          gitlab: {
            project_id: {
              _eq: '29540798'
           }
        }
    }
  }
}
```
Unfortunately, this didn't worked out and it throws the error:  "\"interactors_interactor.gitlab\" is not a relational field"

**Reply by rijkvanzanten** [reply]
> How can we solve this error?

The many to any uses a special syntax to choose what nested collection in the m2a to filter on, see https://docs.directus.io/reference/api/query/#many-to-any-union-types

As for your gitlab example, I believe it's 

```
filter:{
 second_interactor: {
    interactors_id: {
       "interactor:gitlab": {
          project_id: {
            _eq: '29540798'
        }
    }
  }
}
```

**Reply by Disura-Randunu** [reply]
Still no luck. I tried as you suggested and it throws a different error:

> ### **ER_BAD_FIELD_ERROR: Unknown column 'iemee.project_id' in 'where clause**

SQL Query which gets created by Directus:

```
SELECT
    `relationships`.`id`,
    `relationships`.`valid_from`,
    `relationships`.`valid_to`
FROM
    `relationships`
WHERE
    `relationships`.`id` IN(
    SELECT
        `relationships_interactors_1`.`relationships_id` AS `relationships_id`
    FROM
        `relationships_interactors_1`
    LEFT JOIN `interactors` AS `nlkvk`
    ON
        `relationships_interactors_1`.`interactors_id` = `nlkvk`.`id`
    LEFT JOIN `interactors_interactor` AS `iemee`
    ON
        `nlkvk`.`id` = `iemee`.`interactors_id`
    WHERE
        `iemee`.`project_id` = 29540798
)
ORDER BY
    `relationships`.`id` ASC
LIMIT 100

```

@rijkvanzanten, By the way, the table I'm querying is called "relationships" and has several **Many-to-Many** fields. This "second_interactor" field is one of them. It points to the "interactors" table which has a **Many-To-Any** field called 'interactor'. And that field points to several tables (including 'gitlab' table) since it's a **Many-to-Any** field. Does the many-to-many("second_interactor" field) relationship cause some unknown issue?

**Reply by rijkvanzanten** [reply]
> Does the many-to-many("second_interactor" field) relationship cause some unknown issue?

Could be, or it could simply be that my example above doesn't use the right field names for your exact schema. I don't know your exact schema and it's relationships to eachother, which makes it a little difficult to whip out an example like that and have it be perfect first try 🙂

**Comment by sharkfin009** [comment]
@rijkvanzanten hi Rijk! The above docs link is out of date, and I'm struggling to set up permissions for a user to see  the M2A related items on a parent collection .
The docs for filtering relations mention M2M but not M2A.
Thanks so much for any help;)
Ben

**Reply by sharkfin009** [reply]
I can see the problem more clearly now. I need to control access to the related items based on whether they are linked via a junction table to a certain parent item. But there is no field on those items that points to the junction table... this would be the case for M2M or M2A. How would I do this?

**Reply by sharkfin009** [reply]
mmm ... I need to change the database structure to make this work , I see now...

**Comment by martinsotirov** [comment]
@sharkfin009 did you manage to figure this out? I'm struggling with this too.
